### DISCUSSION TITLE
PinClicked Event for Avalonia?

### DISCUSSION THREAD

**Discussion by XDgl233** [Mapsui/Mapsui#2427] [Q&A] [answered]
Hello! I am starting a tech demo using Mapsui and Avalonia for a desktop application that needs to display some offline maps. So far its a charm to work with Mapsui :) But eventually, I came across the problem of how to implement Pin click events. I can draw Pin marker on map control no problem, but it seems there is no PinClicked event like the cases in Mapsui.Forms. Any help is greatly appreciated and thanks in advance :)

**Comment by pauldendulk** [comment, marked as answer]
You should listen to the Map.Info event. You could look at the "Single Callout" demo under "Info".

**Reply by XDgl233** [reply]
Thank you so much :) It works.

**Reply by XDgl233** [reply]
By the way, this may be a very short and quick question, how can I change the font of the widgets? Calloutstyle has font properties, but I can't find related properties in widgets, and when I tried to implement my custom widget, the output is just a white screen, is it something related to the widget renderer? ty :)

**Reply by pauldendulk** [reply]
A CustomWidget also needs a CustomWidgetRenderer that you need to register. Take a look at the custom widget sample and take a look at how that custom renderer is registered in our source code.

I don't know for sure but I expect there is no way to set the font on the widgets, except maybe on specific widgets.

**Reply by pauldendulk** [reply]
Something else. This kind of thing should be on the first page of our docs.

**Reply by pauldendulk** [reply]
I introduced the 'docs-need-update' tag and added it to this discussion. The idea is that on some point later in time we could go through these tags and update the docs for a bunch of these tags.

**Reply by XDgl233** [reply]
Thank you for the reply, I will go take a look at the renderer and looking forward to the new doc :)

**Reply by BasilOmsha** [reply]
I know this is old but can someone possibly tell me how can I draw a pin on the map once the map is clicked?

I tried this so far but to no avail:

```
public void DrawMap()
{
    var map = new Map();
    MapView.Map.Navigator.RotationLock = false;
    MapView.UnSnapRotationDegrees = 30;
    MapView.ReSnapRotationDegrees = 5;


    _myLocationLayer?.Dispose();
    _myLocationLayer = new MyLocationLayer(map)
    {
        IsCentered = false,
    };
    //var location = new Location(location.Longitude, location.Latitude);
    //Debug.WriteLine($" {location.Longitude} {location.Latitude}");

    map.Layers.Add(OpenStreetMap.CreateTileLayer());
    map.Layers.Add(_myLocationLayer);

    //map.Navigator.ZoomToLevel(14);
    //var center = new MPoint(24.478288503698295, 60.97635155698171);
    //var center = new MPoint(location.Longitude, location.Latitude);
    var center = new MPoint(location.Latitude, location.Longitude);
    Debug.WriteLine($" {center.X} {center.Y}");

    MapView.GetMapInfo(center, 5);

    // OSM uses spherical mercator coordinates. So transform the lon lat coordinates to spherical mercator
    var sphericalMercatorCoordinate = SphericalMercator.FromLonLat(center.X, center.Y).ToMPoint();
    // Set the center of the viewport to the coordinate. The UI will refresh automatically
    // Additionally you might want to set the resolution, this could depend on your specific purpose

    map.Home = n => n.CenterOnAndZoomTo(sphericalMercatorCoordinate, n.Resolutions[5]);
    MapView.Map = map;
    _myLocationLayer.UpdateMyLocation(sphericalMercatorCoordinate, true);

    MapView.Info += OnMapClicked;

}

public void OnMapClicked(object sender, MapInfoEventArgs e)
{


    // Get the clicked position
    var clickedPosition = e.MapInfo.WorldPosition;
    Debug.WriteLine($" {clickedPosition.X} {clickedPosition.Y}");



    // Create a pin control
    var pinControl = new PinControl();

     // Create an instance of a Canvas
     var canvas = new Canvas();

    canvas.Children.Add(pinControl);

    // Set the position of the pin control
    Canvas.SetLeft(pinControl, clickedPosition.X - pinControl.Width / 2);
    Canvas.SetTop(pinControl, clickedPosition.Y - pinControl.Height);

   MapView.Content = canvas;

   MapView.RefreshGraphics();;
    
}
```

```
using Avalonia;
using Avalonia.Controls;
using Avalonia.Media;


namespace incomeExpensTrckAVALONIA.Helpers
{
    public class PinControl : Control
    {
        public PinControl()
        {
            Width = 20;
            Height = 20;
        }

        public override void Render(DrawingContext context)
        {
            var rect = new Rect(Bounds.Size);
            var brush = new SolidColorBrush(Colors.Red);

            context.DrawEllipse(brush, null, rect.Center, rect.Width / 2, rect.Height / 2);
        }
    }
}
```

In Maui I was able to do:
```
 public void AddPin(double latitude, double longitude, Color c)
 {
     var myPin = new Pin(MapView)
     {
         Position = new Position(latitude, longitude),
         Type = PinType.Pin,
         Label = "some text",
         Address = "more text",
         Scale = 0.7F,
         Color = c,
     };
     MapView.Pins.Add(myPin);
     PinStored.Add(myPin);
 }
 ```

**Reply by charlenni** [reply]
In MapView, a Layer for the Pins is created for you. If you use a MapControl, and you do this in Abalonia, then you have to create a Layer by your own. In this case, best would be to use a MemoryLayer for this. There you could add your pins.

Look into the "Add Pin" sample of Mapsuis source code.

**Reply by BasilOmsha** [reply]
That is the issue, mapsui Avalonia only has a `MapControl `with no `MapView `that's why I can't use Pins  feature. Would you please give me a heads up or information to how I can make my own layer? Also where can I find info about `MemoryLayer`?

**Reply by charlenni** [reply]
Did you look into the "Add Pin" sample?

**Reply by BasilOmsha** [reply]
Where can I find it?

**Reply by charlenni** [reply]
Samples/Mapsui.Samples.Common/Maps/Demo/WriteToLayerSample.cs

**Reply by BasilOmsha** [reply]
Thanks alot!
